crates/ty_python_semantic/src/types/tuple.rs

```diff
@@ -0,0 +1,819 @@
+        match self {
+            Tuple::Fixed(tuple) => Tuple::Fixed(tuple.normalized(db)),
+            Tuple::Variable(tuple) => Tuple::Variable(tuple.normalized(db)),
+        }
+    }
+
+    fn materialize(&self, db: &'db dyn Db, variance: TypeVarVariance) -> Self {
+        match self {
+            Tuple::Fixed(tuple) => Tuple::Fixed(tuple.materialize(db, variance)),
+            Tuple::Variable(tuple) => Tuple::Variable(tuple.materialize(db, variance)),
+        }
+    }
+
+    fn apply_type_mapping<'a>(&self, db: &'db dyn Db, type_mapping: &TypeMapping<'a, 'db>) -> Self {
+        match self {
+            Tuple::Fixed(tuple) => Tuple::Fixed(tuple.apply_type_mapping(db, type_mapping)),
+            Tuple::Variable(tuple) => Tuple::Variable(tuple.apply_type_mapping(db, type_mapping)),
+        }
+    }
+
+    fn find_legacy_typevars(
+        &self,
+        db: &'db dyn Db,
+        typevars: &mut FxOrderSet<TypeVarInstance<'db>>,
+    ) {
+        match self {
+            Tuple::Fixed(tuple) => tuple.find_legacy_typevars(db, typevars),
+            Tuple::Variable(tuple) => tuple.find_legacy_typevars(db, typevars),
+        }
+    }
+
+    fn has_relation_to(&self, db: &'db dyn Db, other: &Self, relation: TypeRelation) -> bool {
+        match self {
+            Tuple::Fixed(self_tuple) => self_tuple.has_relation_to(db, other, relation),
+            Tuple::Variable(self_tuple) => self_tuple.has_relation_to(db, other, relation),
+        }
+    }
+
+    fn is_equivalent_to(&self, db: &'db dyn Db, other: &Self) -> bool {
+        match (self, other) {
+            (Tuple::Fixed(self_tuple), Tuple::Fixed(other_tuple)) => {
+                self_tuple.is_equivalent_to(db, other_tuple)
+            }
+            (Tuple::Variable(self_tuple), Tuple::Variable(other_tuple)) => {
+                self_tuple.is_equivalent_to(db, other_tuple)
+            }
+            (Tuple::Fixed(_), Tuple::Variable(_)) | (Tuple::Variable(_), Tuple::Fixed(_)) => false,
+        }
+    }
+
+    fn is_gradual_equivalent_to(&self, db: &'db dyn Db, other: &Tuple<'db>) -> bool {
+        match (self, other) {
+            (Tuple::Fixed(self_tuple), Tuple::Fixed(other_tuple)) => {
+                self_tuple.is_gradual_equivalent_to(db, other_tuple)
+            }
+            (Tuple::Variable(self_tuple), Tuple::Variable(other_tuple)) => {
+                self_tuple.is_gradual_equivalent_to(db, other_tuple)
+            }
+            (Tuple::Fixed(_), Tuple::Variable(_)) | (Tuple::Variable(_), Tuple::Fixed(_)) => false,
+        }
+    }
+
+    fn is_disjoint_from(&self, db: &'db dyn Db, other: &Self) -> bool {
+        match (self, other) {
+            (Tuple::Fixed(self_tuple), Tuple::Fixed(other_tuple)) => {
+                self_tuple.is_disjoint_from(db, other_tuple)
+            }
+            // TODO: Consider checking for disjointness between the tuples' prefixes and suffixes.
+            (Tuple::Variable(_), Tuple::Variable(_)) => false,
```

Comment:
The invariant here seems to be that two "pure homogeneous tuples" `tuple[A, ...]` and `tuple[B, ...]` (both tuples without prefixes and suffixes) can never be disjoint even if `A` and `B` are disjoint, because `tuple[Never, ...]` (== `tuple[()]`) would be assignable to both? That might be useful to add as a comment!

Reply:
I _think_ that my reasoning here nonetheless holds (for pure-homogeneous tuples, but not for mixed tuples), so you could consider adding it as a comment 😆